### PR TITLE
Tests: Add missing tests for assertions

### PR DESCRIPTION
**Pull request by leobalter**
Also test failed assertions by mocking `QUnit.assert.push`

These changes were also good to fix assert.throws behaviour.

The other way to test failing assertions I have in mind is reported at #765, but I believe this is the best way to go.

Ref #765 (which might be closed after landing this patch).

### CODE REVIEW

**Comment by leobalter** [comment]
@arschmitz already did something [really similar](https://gist.github.com/arschmitz/9bdcf8065c25ece5748b) on the failing assertions a few days ago. I didn't see that before and it's a crazy coincidence it happened so close.

It might be good to document this as a suggested pattern to test custom assertions.

**Comment by arschmitz** [comment]
+1 for documenting this it took me quite a while to come up with this for my tests on qunit-assert-classes.

File: test/assert.js [hunk 1]
```diff
@@ -237,3 +283,129 @@ QUnit.test( "raises, alias for throws", function( assert ) {
...
 	});
 });
+
+QUnit.module( "failing assertions", {
+	beforeEach: function( assert ) {
+		var originalPush = QUnit.assert.push;
+
+		this.originalPush = originalPush;
+
+		QUnit.assert.push = function( result, actual, expected, message ) {
```
Comment (jzaefferer) [inline test/assert.js]: This should use `assert.push`, that way you're only modifying the current assert instance without requiring access to the global one. May remove the need for the `afterEach` below.

File: src/assert.js [hunk 2]
```diff
@@ -165,11 +165,9 @@ QUnit.assert = Assert.prototype = {
 				expectedOutput = null;
 				ok = true;
 			}
-
-			currentTest.assert.push( ok, actual, expectedOutput, message );
-		} else {
-			currentTest.pushFailure( message, null, "No exception was thrown." );
 		}
+
+		currentTest.assert.push( ok, actual, expectedOutput, message );
```
Comment (jzaefferer) [inline src/assert.js]: This change belongs in a separate commit. Currently its not at all mentioned in the commit message.

Comment (leobalter) [inline src/assert.js]: Done

File: test/assert.js [hunk 3]
```diff
@@ -237,3 +283,122 @@ QUnit.test( "raises, alias for throws", function( assert ) {
 		throw "my error";
 	});
 });
+
+QUnit.module( "failing assertions", {
+	beforeEach: function( assert ) {
+		assert.push = function( result, actual, expected, message ) {
+
+			// inverts the result so we can test failing assertions
+			QUnit.assert.push.call( assert, !result, actual, expected, message );
```
Comment (jzaefferer) [inline test/assert.js]: Would be good to get rid of this global as well, instead saving a reference to the original `assert.push` and calling the original here.

Comment (leobalter) [inline test/assert.js]: As the QUnit.assert is a reference for Assert's prototype and we can just call it with assert, I believe it's better than storing a variable in the memory, also it takes less loc.

Comment (jzaefferer) [inline test/assert.js]: I'd like to get rid of all references to `QUnit.assert` outside of QUnit's source code, except those adding custom assertions.

In this case it should also remove the need to use `call()`.

**Comment by leobalter** [comment]
done

**Comment by jzaefferer** [comment]
Looks good to land!

**Comment by leobalter** [comment]
Thanks!
